src/commonMain/kotlin/com/mattprecious/stacker/stack/tree.kt

```diff
@@ -0,0 +1,81 @@
+package com.mattprecious.stacker.stack
+
+fun <K : Any, T : Any> treeOf(
+	elements: Collection<T>,
+	keySelector: (T) -> K,
+	parentSelector: (T) -> K?,
+): TreeNode<T>? {
+	if (elements.isEmpty()) {
+		return null
+	}
+
+	val elementsByKey = elements.associateBy(keySelector)
+	val parents = elements.associateBy(keySelector) {
```

Comment:
This can be a `mapValues` operation on `elementsByKey` rather than having to regenerate all the keys.